code/espurna/mcp23s17.cpp

```diff
@@ -0,0 +1,222 @@
+
+	pinMode(CS, OUTPUT);
+	digitalWrite(CS, HIGH);
+	MCP23S17InitGPIO();
+}
+
+/**
+ * @brief Set a expander MCP23S17 the pin direction.
+ *
+ * @param pinNumber Pin number for set.
+ * @param mode direction mode. 0 - INPUT, 1 - OUTPUT.
+ *
+ * @return void
+ */
+void MCP23S17InitGPIO()
+{
+	// Relays.
+	for (uint8_t i = 0; i < DUMMY_RELAY_COUNT; i++)
+	{
+		DEBUG_MSG_P(PSTR("[MCP23S17] Initialize GPIO %d\n"), i);
+		MCP23S17SetDirection(RELAY_PINS[i], OUTPUT);
+	}
+
+	// Opto inputs.
+	for (uint8_t i = 0; i < OPTOIN_COUNT; i++)
+	{
+		MCP23S17SetDirection(OPTOIN_PINS[i], INPUT);
+	}
+}
+
+/**
+ * @brief Set a expander MCP23S17 the pin direction.
+ *
+ * @param pinNumber Pin number for set.
+ * @param mode direction mode. 0 - INPUT, 1 - OUTPUT.
+ *
+ * @return void
+ */
+void MCP23S17SetDirection(uint8_t pinNumber, uint8_t mode)
+{
+	uint8_t registerData = MCP23S17ReadRegister(IODIR);
+
+	if (INPUT == mode)
+	{
+		registerData |= (1 << pinNumber);
+	}
+	else
+	{
+		registerData &= ~(1 << pinNumber);
+	}
+
+	MCP23S17WriteRegister(IODIR, registerData);
+}
+ 
+/**
+ * @brief Read an expander MCP23S17 a register.
+ *
+ * @param address A register address.
+ *
+ * @return The data from the register.
+ */
+uint8_t MCP23S17ReadRegister(uint8_t address)
+{
+	_expTxData[0] = READ_CMD;
+	_expTxData[1] = address;
+
+	digitalWrite(CS, LOW);
+	SPI.transferBytes(_expTxData, _expRxData, 3);
+	digitalWrite(CS, HIGH);
+
+	return _expRxData[2];
+} 
+
+/**
+ * @brief Write data in expander MCP23S17 register.
+ *
+ * @param address A register address.
+ * @param data A byte for write.
+ *
+ * @return void.
+ */
+void MCP23S17WriteRegister(uint8_t address, uint8_t data)
+{
+	_expTxData[0] = WRITE_CMD;
+	_expTxData[1] = address;
+	_expTxData[2] = data;
+
+	digitalWrite(CS, LOW);
+	SPI.transferBytes(_expTxData, _expRxData, 3);
+	digitalWrite(CS, HIGH);
+} 
+
+/**
+ * @brief Set expander MCP23S17 pin state.
+ *
+ * @param pinNumber The number of pin to be set.
+ * @param state The pin state, true - 1, false - 0.
+ *
+ * @return void
+ */
+void MCP23S17SetPin(uint8_t pinNumber, bool state)
+{
+	uint8_t registerData = MCP23S17ReadRegister(OLAT);
+
+	if (state)
+	{
+		registerData |= (1 << pinNumber);
+	}
+	else
+	{
+		registerData &= ~(1 << pinNumber);
+	}
+
+	MCP23S17WriteRegister(OLAT, registerData);
+}
+
+/**
+ * @brief Get MCP23S17 MCP23S17 pin state.
+ *
+ * @param pinNumber The number of pin to be get.
+ *
+ * @return State true - 1, false - 0.
+ */
+bool MCP23S17GetPin(uint8_t pinNumber)
+{
+	uint8_t registerData = MCP23S17ReadRegister(GPIO);
+
+	return registerData & (1 << pinNumber);
+}
+ 
+/**
+ * @brief Set relay new state.
+ *
+ * @param relayNumber Number of relay from 0 to RELAY_COUNT - 1. 0 - Relay1, 1 - Relay2 ...
+ * @param state New state of relay, true - On, false = Off.
+ *
+ * @return void
+ */
+void MCP23S17SetRelayState(uint8_t relayNumber, bool state)
+{
+	// Check if relayNumber is out of range - return.
+	if (relayNumber > RELAY_COUNT - 1)
```

Comment:
`relayCount()`?

Reply:
changed as DUMMY_RELAY_COUNT

(i don't want to import relay.h)

Reply:
Ah, yes, no need for relay-relay here. This counts as a module setting, my mistake.